### PR TITLE
feat(python): improve `read_excel` table data identification

### LINKED ISSUE #10831: "found more fields than defined in 'Schema'" error when using read_excel
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python

pl.read_excel("limits.xlsx", sheet_id=0)

```
[limits.xlsx](https://github.com/pola-rs/polars/files/12487153/limits.xlsx)

### Issue description

With versions prior to 0.19.0 my excel file was readable even if the sheets had different unused extra columns.

With version 0.19.0 this does not work for me any more.

### Expected behavior

I would have expected that the missing columns would become missing columns in the dataframe.

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.0
Index type:          UInt32
Platform:            Windows-10-10.0.19041-SP0
Python:              3.8.0 (tags/v3.8.0:fa919fd, Oct 14 2019, 19:37:50) [MSC v.1916 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
matplotlib:          <not installed>
numpy:               <not installed>
pandas:              <not installed>
pyarrow:             <not installed>
pydantic:            <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            0.8.1
xlsxwriter:          3.1.2
```

</details>

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes #10831.

This adds several per-engine improvements for identifying where tables begin/end (with the primary aim of automatically omitting all-null unnamed columns that are incorrectly imported due to worksheet used-range inference).

* `xlsx2csv` _(default)_
  Set "truncate_ragged_lines" to True by default. There appears to be no downside here; if it is not necessary then there is no change in behaviour - if it is, we don't force the user to set it to avoid the "found more fields than defined in 'Schema'" error. 

* `openpyxl` _(added earlier today)_
  This engine is able to exactly detect "real" Excel Table objects (the same thing we output from `write_excel`), so we favour reading the data from the range described by the table ref (omitting any dynamic "Totals" row) if we detect this. This _guarantees_ that we will read the table data range accurately.
  
  If reading _non_-Table sheet data, we only start ingesting rows in the used range that are not all-null. 

The DataFrame produced from both engines will now be checked before return, and unnamed all-null columns will be dropped.

It's not possible to determine whether trailing all-null _rows_ are valid or not. Might be worth adding an additional parameter so the caller can opt-in to this as a convenience; can add if wanted.

Added two additional sheets to the `tests/unit/io/files/example.xlsx` workbook, both containing tables in a larger used-range that do not start in the top-left corner (one sheet containing raw data, the other containing a "real" Table object) and used these to extend test coverage and validate the new behaviour for both engines.

---

**Also**: while I was at it, I extended the `sheet_id` and `sheet_names` params to support taking multiple items. We can already load _all_ sheets into a `{sheetname:frame, ...}` dict by setting sheet id = 0, and this allows the same for _selected_ sheets.

### CODE REVIEW

File: py-polars/polars/io/excel/functions.py [hunk 1]
```diff
@@ -251,6 +282,19 @@ def read_excel(
         )
 
 
+def _drop_null_columns(df: pl.DataFrame) -> pl.DataFrame:
+    # drop all-null cols with no name
```
Comment (stinodego) [inline py-polars/polars/io/excel/functions.py:286]: I'd say either make this a docstring, or even better: rename the function to `_drop_unnamed_null_columns` and you don't need a comment at all.

File: py-polars/polars/io/excel/functions.py [hunk 2]
```diff
@@ -196,10 +213,15 @@ def read_excel(
 
     if xlsx2csv_options is None:
         xlsx2csv_options = {}
+
     if read_csv_options is None:
-        read_csv_options = {}
+        read_csv_options = {"truncate_ragged_lines": True}
+    elif "truncate_ragged_lines" not in read_csv_options:
+        read_csv_options["truncate_ragged_lines"] = True
```
Comment (stinodego) [inline py-polars/polars/io/excel/functions.py:220]: Won't this incur a performance penalty?

File: py-polars/tests/unit/io/files/example.xlsx
Comment (stinodego) [inline py-polars/tests/unit/io/files/example.xlsx:1]: Why do we need a file now, when we didn't need one before?

**Review by stinodego — APPROVED** [review]
Thanks! I think this looks fine, just a few comments for clarification.

Also: I think we're getting to the point where this needs to be refactored. Too much chained / nested / if / else stuff going on for my liking. I'll approve this, but I'd appreciate it if you could attempt to untangle some of this before adding more functionality.

**Comment by alexander-beedie** [comment]
> attempt to untangle some of this before adding more functionality.

We're on the same page - you should have seen this function a few weeks ago, heh. I've already done one substantial despaghettification here, and I'll see about doing another round  ;)

File: py-polars/tests/unit/io/files/example.xlsx
Comment (alexander-beedie) [reply py-polars/tests/unit/io/files/example.xlsx:1]: This file already existed - I just added two new worksheets to it so we could validate the additional functionality  ;)

File: py-polars/polars/io/excel/functions.py (hunk 1 above)
Comment (alexander-beedie) [reply py-polars/polars/io/excel/functions.py]: Done 👍

**Comment by alexander-beedie** [comment]
>  I think we're getting to the point where this needs to be refactored.

Just did another pass to factor-out the engine/parser init and tidy up the main function body.

File: py-polars/polars/io/excel/functions.py (hunk 2 above)
Comment (alexander-beedie) [reply py-polars/polars/io/excel/functions.py:216]: Fortunately not - it's close to zero cost here, which is why I don't feel bad enabling it by default ;)

**Pull request merged**